File: biobakery_workflows/scripts/merge_paired_ends.R
#!/usr/bin/env Rscript

# load packages
library(dada2); packageVersion("dada2")
library(tools)

args <- commandArgs(TRUE)

## Parse arguments (we expect the form --arg=value)
parseArgs <- function(x) strsplit(sub("^--", "", x), "=")
args.df <- as.data.frame(do.call("rbind", parseArgs(args)))

args.list <- as.list(as.character(args.df$V2))
names(args.list) <- args.df$V1

## Arg1 default
if(is.null(args.list$output_dir)) {
  stop("At least one argument must be supplied (output folder).\n", call.=FALSE)
}

# Print args list to STDOUT
for( i in names(args.list) ) {
  cat( i, "\t", args.list[[i]], "\n")
}

# print contents of folder
cat( grep( "*\\.fastq*", list.files(args.list$output_dir), value=T ), sep = "\n" )

# these variables are passed to the workflow
output.path <- normalizePath(args.list$output_dir )
print(output.path)

#Filtered files folder path
filt_path <- file.path(output.path, "filtered_input") 

# Sort ensures forward/reverse reads are in same order
fnFs <- sort(grep( "_F_filt.*\\.fastq", list.files(filt_path), value = T ) )
fnRs <- sort(grep( "_R_filt.*\\.fastq", list.files(filt_path), value = T ) )


# Extract sample names, allowing variable filenames; e.g. *_R1[_001].fastq[.gz]
sample.names <- gsub( "_F_filt.*\\.fastq", "", fnFs, perl = T)
sample.namesR <- gsub( "_R_filt.*\\.fastq", "", fnRs, perl = T)

sample.ext <- tools::file_ext(fnFs)
if(identical("gz",sample.ext[1])){
  sample.ext <- "fastq.gz"
  # Extract sample names, allowing variable filenames
  sample.names <- gsub( "_F_filt.*\\.fastq.gz", "", fnFs, perl = T)
  sample.namesR <- gsub( "_R_filt.*\\.fastq.gz", "", fnRs, perl = T)
}


if(!identical(sample.names, sample.namesR)) stop("Forward and reverse files do not match.")

cwd <- getwd()

# Define filenames for filtered input files
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.", sample.ext))
filtRs <- file.path(filt_path, paste0(sample.names, "_R_filt.", sample.ext))

errF <- readRDS(file.path(output.path,"error_rates_F.rds"))
errR <- readRDS(file.path(output.path,"error_rates_R.rds"))


# Sample inference of dereplicated reads, and merger of paired-end reads
mergers <- vector("list", length(sample.names))
names(mergers) <- sample.names
names(filtFs) <- sample.names
names(filtRs) <- sample.names
for(sam in sample.names) {
  cat("Processing:", sam, "\n")
  print(filtFs[[sam]])
  derepF <- dada2::derepFastq(filtFs[[sam]])
  ddF <- dada2::dada(derepF, err=errF, multithread=TRUE)
  derepR <- dada2::derepFastq(filtRs[[sam]])
  ddR <- dada2::dada(derepR, err=errR, multithread=TRUE)
  merger <- dada2::mergePairs(ddF, derepF, ddR, derepR)
  mergers[[sam]] <- merger
}
rm(derepF); rm(derepR)
saveRDS(mergers, paste0(output.path, "/mergers.rds"))

File: biobakery_workflows/scripts/merge_paired_ends.py
#!/usr/bin/env python

from rpy2 import robjects
from rpy2.robjects.packages import importr
import os.path
import sys
import re
import argparse

base = importr('base')
dada2 = importr('dada2')

def parse_arguments(args):
    """ Parse the arguments"""

    parser = argparse.ArgumentParser(
        description="Merge paired ends arguments\n",
        formatter_class=argparse.RawTextHelpFormatter)

    parser.add_argument(
        "--filt_path",
        help="path to filtered files \n[REQUIRED]",
    )
    parser.add_argument(
        "--error_ratesF_rds",
        help="error rates forward reads data\n[REQUIRED]",
    )
    parser.add_argument(
        "--error_ratesR_rds",
        help="error rates reverse reads data\n[REQUIRED]",
    )
    parser.add_argument(
        "--mergers_rds",
        help="merged reads rds file\n[REQUIRED]",
    )
    parser.add_argument(
        "--sample_names_rds",
        help="sample names rds file\n[REQUIRED]",
    )
    parser.add_argument(
        "--threads",
        help="number of threads\n[REQUIRED]",
    )

    return parser.parse_args()


def main():

    # parse arguments from the user
    args = parse_arguments(sys.argv)

    filtFs = sorted([f for f in os.listdir(args.filt_path) if re.match(r'(.*)_F_filt(.*)', f)])
    filtRs = sorted([f for f in os.listdir(args.filt_path) if re.match(r'(.*)_R_filt(.*)', f)])

    sample_names = [f.split("_F_filt")[0] for f in filtFs]
    sample_namesR = [f.split("_R_filt")[0] for f in filtRs]

    print(sample_names)

    if (sample_names != sample_namesR):
        sys.exit("ERROR: Forward and reverse files do not match.")

    filtFs = [os.path.join(args.filt_path, f) for f in filtFs]
    filtRs = [os.path.join(args.filt_path, f) for f in filtRs]

    robjects.globalenv["error_ratesF_rds"] = args.error_ratesF_rds
    robjects.globalenv["error_ratesR_rds"] = args.error_ratesR_rds
    robjects.globalenv["mergers_rds"] = args.mergers_rds
    robjects.globalenv["filtFs"] = filtFs
    robjects.globalenv["filtRs"] = filtRs
    robjects.globalenv["sample.names"] = sample_names
    robjects.globalenv["sample_names_rds"] = args.sample_names_rds
    robjects.globalenv["threads"] = args.threads

    robjects.r('''
      
        errF <- readRDS(error_ratesF_rds)
        errR <- readRDS(error_ratesR_rds)
        
        filtF <- unlist(filtFs, use.names=FALSE)
        filtR <- unlist(filtRs, use.names=FALSE)
        sample.names <- unlist(sample.names, use.names=FALSE)
        saveRDS(sample.names, sample_names_rds)
        mergers <- vector("list", length(sample.names))
        names(mergers) <- sample.names
        names(filtF) <- sample.names
        names(filtR) <- sample.names
        for(sam in sample.names) {
          cat("Processing:", sam, "\n")
          print(filtF[[sam]])
          derepF <- dada2::derepFastq(filtF[[sam]])
          ddF <- dada2::dada(derepF, err=errF, multithread=as.numeric(threads))
          derepR <- dada2::derepFastq(filtR[[sam]])
          ddR <- dada2::dada(derepR, err=errR, multithread=as.numeric(threads))
          merger <- dada2::mergePairs(ddF, derepF, ddR, derepR)
          mergers[[sam]] <- merger
        }
        rm(derepF); rm(derepR)
        
        saveRDS(mergers, mergers_rds)
    
        ''')

if __name__ == "__main__":
        main()
